File: Server/Tuckshop.Domain/Tuckshop.Core.Models/Models/Orders/Queries/OrderLookup.cs
#pragma warning disable
namespace Tuckshop.Models.Orders.Queries
{
  using System;
  using System.Collections.Generic;
  using System.Linq;
  using Tuckshop.Core.Models;

  public class OrderLookup
  {
    public int OrderId { get; set; }
    public string CustomerName { get; set; }
    public DateTime OrderedOn { get; set; }
    public DateTime? CompletedOn { get; set; }
    public DateTime? CancelledOn { get; set; }
    public string CancelledReason { get; set; }
    public string CompletedBy { get; set; }
    public string CancelledBy { get; set; }
    public decimal OrderTotalExcl { get; set; }
    public decimal OrderTotal { get; set; }
    public List<OrderDetailLookup> Items { get; set; }

    /// <summary>
    /// Sets the <see cref="Items"/> property, and returns the order.
    /// </summary>
    /// <param name="orderDetails">Order details.</param>
    /// <returns>This instance.</returns>
    public OrderLookup WithDetails(IEnumerable<OrderDetailLookup> orderDetails)
    {
      this.OrderTotal = orderDetails.Sum(c => c.Value);
      this.OrderTotalExcl = orderDetails.Sum(c => c.Value - c.VAT);
      this.Items = orderDetails.ToList();
      return this;
    }
  }
}

File: Client/Tuckshop.WebApp/src/Domain/Models/Orders/Queries/OrderLookup.ts
import { Attributes, List, LookupBase } from '@singularsystems/neo-core';
import OrderDetail from '../OrderDetail';
import OrderDetailLookup from './OrderDetailLookup';

export default class OrderLookup extends LookupBase {

    constructor() {
        super();
        this.makeBindable();
    }

    public readonly orderId: number = 0;

    public readonly customerName: string = "";

    @Attributes.Date()
    public readonly orderedOn: Date = new Date();

    @Attributes.Observable()
    @Attributes.Date()
    public completedOn: Date | null = null;

    @Attributes.Observable()
    @Attributes.Date()
    public cancelledOn: Date | null = null;

    public cancelledReason: string = "";

    public readonly completedBy: string = "";

    public readonly cancelledBy: string = "";

    @Attributes.Float()
    public readonly orderTotalExcl: number = 0;

    @Attributes.Float()
    public readonly orderTotal: number = 0;

    public readonly items = new List(OrderDetailLookup);

    // Client only properties / methods
    @Attributes.Observable()
    public isExpanded = false;

    public get canAction() {
        return !this.completedOn && !this.cancelledOn;
    }
}